src/tests/unittests/asn1.c

```diff
@@ -0,0 +1,379 @@
+	u8 data3[] = {0x55, 0x04, 0x87, 0xFF, 0xFF, 0xFF, 0x7F};
+	/* 2.2147483647.4.3 (The second part is largest 32 bit integer) */
+	u8 data4[] = {0x88, 0x80, 0x80, 0x80, 0x4F, 0x04, 0x03};
+	/* 2.5.4.2147483648 (The last part is 32 bit integer overflow) */
+	u8 data5[] = {0x55, 0x04, 0x88, 0x80, 0x80, 0x80, 0x00};
+	/* 2.2147483648.4.3 (The second part is 32 bit integer overflow) */
+	u8 data6[] = {0x88, 0x80, 0x80, 0x80, 0x50, 0x04, 0x03};
+	struct sc_object_id oid;
+	u8 *buf = NULL;
+	size_t buflen = 0;
+	int rv = 0;
+
+	rv = sc_asn1_decode_object_id(data1, sizeof(data1), &oid);
+	assert_int_equal(rv, SC_ERROR_NOT_SUPPORTED);
+
+	rv = sc_asn1_decode_object_id(data2, sizeof(data2), &oid);
+	assert_int_equal(rv, SC_ERROR_NOT_SUPPORTED);
+
+	if (sizeof(int) == 4) {
+		/* For 64 bit builds */
+		struct sc_object_id oid_3 = {{2, 5, 4, 2147483647, -1}};
+		struct sc_object_id oid_4 = {{2, 2147483647, 4, 3, -1}};
+
+		rv = sc_asn1_decode_object_id(data3, sizeof(data3), &oid);
+		assert_int_equal(rv, SC_SUCCESS);
+		assert_int_equal(sc_compare_oid(&oid_3, &oid), 1);
+		rv = sc_asn1_encode_object_id(&buf, &buflen, &oid);
+		assert_int_equal(rv, SC_SUCCESS);
+		assert_int_equal(buflen, sizeof(data3));
+		assert_memory_equal(buf, data3, buflen);
+		free(buf);
+
+		rv = sc_asn1_decode_object_id(data4, sizeof(data4), &oid);
+		assert_int_equal(rv, SC_SUCCESS);
+		assert_int_equal(sc_compare_oid(&oid_4, &oid), 1);
+		rv = sc_asn1_encode_object_id(&buf, &buflen, &oid);
+		assert_int_equal(rv, SC_SUCCESS);
+		assert_int_equal(buflen, sizeof(data4));
+		assert_memory_equal(buf, data4, buflen);
+		free(buf);
+
+		rv = sc_asn1_decode_object_id(data5, sizeof(data5), &oid);
+		assert_int_equal(rv, SC_ERROR_NOT_SUPPORTED);
+
+		rv = sc_asn1_decode_object_id(data6, sizeof(data6), &oid);
+		assert_int_equal(rv, SC_ERROR_NOT_SUPPORTED);
+	} else {
+		rv = sc_asn1_decode_object_id(data3, sizeof(data3), &oid);
+		assert_int_equal(rv, SC_ERROR_NOT_SUPPORTED);
+
+		rv = sc_asn1_decode_object_id(data4, sizeof(data4), &oid);
+		assert_int_equal(rv, SC_ERROR_NOT_SUPPORTED);
+	}
+}
+
+static void torture_oid(void **state)
+{
+	/* (without the tag and length {0x06, 0x06}) */
+	/* Small OIDs */
+	struct sc_object_id small_oid = {{0, 1, 2, 3, 4, 5, 6, -1}};
+	u8 small[] = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06};
+	/* Limit what we can fit into the first byte */
+	struct sc_object_id limit_oid = {{2, 47, -1}};
+	u8 limit[] = {0x7F};
+	/* The second octet already oveflows to the second byte */
+	struct sc_object_id two_byte_oid = {{2, 48, -1}};
+	u8 two_byte[] = {0x81, 0x00};
+	/* Existing OID ec publickey */
+	struct sc_object_id ecpubkey_oid = {{1, 2, 840, 10045, 2, 1, -1}};
+	u8 ecpubkey[] = {0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x02, 0x01};
+	/* Missing second byte, even though indicated with the first bit */
+	u8 missing[] = {0x81};
+	/* Missing second byte in later identifiers */
+	u8 missing_second[] = {0x2A, 0x48, 0x81};
```

Comment:
what about non-minimal encoding of the subidentifier? something like `{0x2a, 0x80, 0x01}` (as an incorrect encoding of (1, 2, 1)) or `{0x80, 0x01}`

Reply:
I will add these